core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseConnectionCache.scala

```
@@ -30,73 +33,91 @@ import org.apache.spark.Logging
 private[spark] object HBaseConnectionCache extends Logging {
 
   // A hashmap of Spark-HBase connections. Key is HBaseConnectionKey.
-  val connectionMap = new mutable.HashMap[HBaseConnectionKey, SmartConnection]()
+  private[hbase] val connectionMap = new mutable.HashMap[HBaseConnectionKey, SmartConnection]()
 
-  val cacheStat = HBaseConnectionCacheStat(0, 0, 0)
+  private val cacheStat = HBaseConnectionCacheStat(0, 0, 0)
 
   // in milliseconds
   private final val DEFAULT_TIME_OUT: Long = SparkHBaseConf.connectionCloseDelay
-  private var timeout = DEFAULT_TIME_OUT
-  private var closed: Boolean = false
+  private val timeout = new AtomicLong(DEFAULT_TIME_OUT)
+  private val closed = new AtomicBoolean(false)
 
-  var housekeepingThread = new Thread(new Runnable {
+  private var housekeepingThread = new Thread(new Runnable {
     override def run() {
       while (true) {
-        try {
-          Thread.sleep(timeout)
-        } catch {
-          case e: InterruptedException =>
-            // setTimeout() and close() may interrupt the sleep and it's safe
-            // to ignore the exception
-        }
-        if (closed) {
-          return
-        }
+        sleep(timeout.get(), allowInterrupt = true, allowClosed = true)
+        if (closed.get()) return
         performHousekeeping(false)
       }
     }
   })
   housekeepingThread.setDaemon(true)
   housekeepingThread.start()
 
+  // Thread.sleep can be spuriously woken up, this ensure we sleep for atleast the
+  // 'duration' specified
+  private[hbase] def sleep(duration: Long, allowInterrupt: Boolean = false, allowClosed: Boolean = false): Unit = {
+    val startTime = System.currentTimeMillis()
+    var remaining = duration
+    while (remaining > 0) {
+      try {
+        Thread.sleep(remaining)
+      } catch {
+        case ex: InterruptedException if allowInterrupt => return
+        case ex: Exception => // ignore
```

Why ignore the exception ?